schemas/schema-kinds.md

```diff
@@ -1,29 +1,65 @@
-Schema Kinds
-------------
+# Schema Kinds
 
-Recursive types contain additional type definitions for either their
-keys (in the case of maps),
-values (in the case of maps and lists),
-or fields (in the case of structs).
+* [Extending the IPLD Data Model](#extending-the-ipld-data-model)
+  * [Data Model Kinds](#data-model-kinds)
+  * [Recursive Kinds](#recursive-kinds)
+  * [Meta Kinds](#meta-kinds)
+* [Value Type Modifiers](#value-type-modifiers)
+  * [Nullable Values](#nullable-values)
+  * [Optional Fields](#optional-fields)
+  * [Fields with Defaults](#fields-with-defaults)
+  * [Combining Nullable, Optional, and Default](#combining-nullable-optional-and-default)
+* [Understanding Cardinality](#understanding-cardinality)
+  * [Cardinality Examples](#cardinality-examples)
 
-### table
 
-- Null
-- Boolean
-- Integer
-- Float
-- String
-- Bytes
-- List
-- Map
-- Link
-- Union
-- Struct
-- Enum
+## Extending the IPLD Data Model
 
+IPLD Schemas define a set of "kinds" that are built upon the
+[IPLD Data Model](https://github.com/ipld/specs/blob/master/data-model-layer/data-model.md#kinds):
 
-Value Type Modifiers
---------------------
+### Data Model Kinds
+
+  - **Null**
+  - **Boolean**
+  - **Integer**
+  - **Float**
+  - **String**
+  - **Bytes**
+  - **List**
+  - **Map**
+  - **Link**
+
+### Recursive Kinds
```

Comment:
I love the docs and reorg into sections you've done here, but this section name is wrong.  Map and List are recursive kinds, and they're in the Data Model Kinds set; therefore this heading can't be "recursive".

I can think of a couple choices for what else we could title this, and maybe you can brainstorm others as well:

- "Type-level kinds"
- "Perceived kinds" ([ancient reference](https://gist.github.com/warpfork/6df17e791936d1f9b0d5e5483678c8bf#file-gistfile1-txt-L3) -- not attached to it tho)
- ...?

We also might want to decide on if we orient most of our phrasing around the two *sets*... or the DM set plus more.  I tend to favor the former approach (naming the two sets), and then maybe don't spent so much words on the naming the ones that differ; just briefly pointing them out might suffice.

Reply:
Though it touches on things beyond the original scope of this PR, I'll throw on for consideration: maps and lists also behave slightly different in the schema layer, because they have to have specified key and value types now.  (One *can* say `[any]` and `{String:any}`, but one now has to... well, say that.)

So this might be an indication that the two-sets style of description might be the clearer route: if gives a place where this shift in map and list behavior can also be briefly discussed.

Reply:
what about "Abstract Kinds"?

Reply:
I don't quite understand what you mean by the last paragraph, is there something concrete in this PR I could change to alleviate that concern?

Reply:
"Type-level kinds" -> would describe all of {int,string,...,map,list,union,enum,struct}.
"Perceived kinds" -> would describe just {union,enum,struct} -- the added ones.
(I suspect "abstract kinds" would lean towards the latter mode as well.)

I tend to think the former route might be preferable, because the latter route results in docs *everywhere else* having to say the mouthful of "the data model plus perceived kinds".

Reply:
OK, how about this:

-----

**Scalar Kinds**

... something about how this is the data model minus list and map, list them here ...

**Recursive Kinds**

... something like: these kinds build upon the scalars in various ways, maps and lists in the schema layer are not entirely freeform and must be composed of specific types, as specified in a schema ... structs, unions and enums are abstract kinds that describe layouts comprised of data model kinds, depending on the representations specified ...

Reply:
and **Meta Kinds** since you want that to be a separate thing, or do you have a better suggestion for how to handle `copy`?

Reply:
> the former route might be preferable

you're losing me, which route? what headings do you want?

Reply:
Okay, lemme back myself up a bit.

Usually when I visualize these things, I visualize them as a venn diagram.  The Data Model Kinds are a subset of the Type Level Kinds.  So we've got a circle inside a bigger circle.  And in my mental model, those are the two labels I use -- I label each of the two big circles.  I usually _don't_ label the circle-with-a-cutout shape (which is where struct,union,enum live) but if I did, I'd probably call it the "perceived kinds".

The reason I mentally label it this way is that I've found I end up needing to refer to "type-level kinds" in other documentation or code comments at a rate of about 10x more often than I need refer to "perceived kinds".

(We've got another wee small circle on the side of the venn diagram now for "meta kinds" for the Copy feature.)

So, it seems to me there's two different ways to write this documentation: naming the ~two~ three big circles; or naming the circle, the circle-with-cutout, and the other wee circle.  (This is the 'route' I was referring to).

In either case, the "meta kinds" header is the same.  And in either case, the reference to the data model kinds is there, so that heading is probably the same.  The only heading that changes based on this choice is the middle one -- because that's the thing that's either a circle, or a circle-with-cutout, in the venn diagram.

So we've got two choices: what to call this header, and which approach to the venn diagram we're taking.  But they're actually the same choice.  Does that savvy?

---

I was considering advocating for the name-the-circles route and thus avoiding having to name the circle-with-cutout, because of the aforementioned ratio in usage rate.  (In that route, we'd probably favor a name more like "type-level kinds" (if not that specific string) for this section -- and then the first sentence in this section would be "The type-level kinds include all of the already familiar data-model kinds, plus three new kinds:" and then resume with the rest of the paragraphs below.  See how the "three new kinds" never got a group title themselves?)

But now that I've backed myself up a bit, I see how describing {circle, circle-with-cutout, circle} is probably more obvious when writing it as a bulleted list.  I'm just a bit sad it means having to name the circle-with-cutout region at all.

This just turned into a bunch of braindump, sorry.  Maybe the one good takeaway is that we should actually consider a venn diagram png somewhere near here!

Reply:
I'm thinking that maybe this doc would be better by:

1. Leading with a listing the data model kinds
2. Then moving into a full list of type kinds with:
  a. An explanation of the new restrictions placed on list and map (although the restrictions are hard to be clear about because unions let you have mixed kinds within lists and map values).
  b. How the new forms, struct, enum and union, build compose from the other kinds and how they can be described as "perceived".
  c. Mention "copy" as a meta kind.